### DISCUSSION TITLE
Only top level suggestions are offered for auto imports

### DISCUSSION THREAD

**Discussion by edraft** [microsoft/pylance-release#2681] [Enhancement]
### VS Code version

1.62.2

### Extension version

v2021.11.1422169775

### OS type

Linux

### OS version

Ubuntu 20.04.3

### Python distribution

python.org

### Python version

3.9.2

### Language server

Pylance

### Expected behaviour

I would expect vs code to suggest imports from venvs.

If you want to install a package on the venv, for example discord.py, and then import, for example, `commands.Bot` from discord.ext, press Ctrl +. to `commands`.
I would expect the import `from discord.ext import commands` to be suggested, but it is not.

### Actual behaviour

Any other import found is suggested, but not from packages in venv.

This also doesn't work with auto-imports.

### Steps to reproduce

```sh
python -m venv venv

source venv/bin/activate

pip install discord.py
```

Create Python file:
```py
commands.Bot.__init__()
```

`commands` is not defined so try to import it...
If it works, the problem is mine

### Logs

I think this line is relevant:

[Info  - 18:57:41] Auto-excluding /home/user/project/venv

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by edraft** [comment]
The extension brainfit.vscode-importmagic can resolve imports correctly.
But its not really good...

**Comment by bschnurr** [comment]
By design we only offer top level suggestions for auto imports, except for a number of popular packages.  This was a perf trade off.   

In the future we could address this issue by making the levels configurable.   For now we will wait for up votes on this issue

**Comment by edraft** [comment]
Would be great to customize :D

**Comment by edraft** [comment]
> By design we only offer top level suggestions for auto imports, except for a number of popular packages. This was a perf trade off.
> 
> In the future we could address this issue by making the levels configurable. For now we will wait for up votes on this issue

Is there any fix/hack for the moment?
Without this feature python in vs code is nearly unusable for me.

**Comment by heejaechang** [comment]
currently there is no workaround.

**Comment by ldorigo** [comment]
Would also like this, I'm often frustrated by auto imports (sometimes they work, sometimes they don't, it feels a bit arbitrary). 
Regarding perf. - maybe you could index the environment in the background to provide auto-import suggestions (I think Pycharm does this)? 

Edit: Ah, apparently you're already on it https://github.com/microsoft/pylance-release/issues/2069 😁

**Comment by cspwizard** [comment]
@bschnurr Hi! Where I can vote for the feature? It's useless when autocomplete/import ignores packages installed in the venv.

**Comment by tkoenig89** [comment]
> By design we only offer top level suggestions for auto imports, except for a number of popular packages. This was a perf trade off.
> 

I think this should be stated more obviously in the readme. I read "auto import" in the readme and spent half an hour trying to 'fix' my VS Code, venv, etc... setup, because i could only get the `gc` or `os` auto import working. But auto importing 'popular' packages like `pandas` or `numpy` did not work (Edit: fixed this issue, using `python.analysis.indexing=true` which is somewhat [hidden](https://github.com/microsoft/pylance-release/issues/28#issuecomment-728219527), so nevermind this part).

It took me a while to get here to find out, that auto import only works on arbitrary packages. As there might be more people like me, you might get less 'bug' reports on this feature, if this is stated more clearly in the readme.

**Comment by bschnurr** [comment]
fyi to override the default indexing add this to your settings.json

```
    "python.analysis.packageIndexDepths":[
        ["", 1],
        ["sklearn", 2],
        ["matplotlib", 2],
        ["scipy", 2],
        ["django", 2],
        ["numpy",2]
    ],
```

**Comment by cspwizard** [comment]
> "python.analysis.packageIndexDepths":[
>         ["", 1],
>         ["sklearn", 2],
>         ["matplotlib", 2],
>         ["scipy", 2],
>         ["django", 2],
>         ["numpy",2]
>     ],

Great! Thanks @bschnurr,  could you  please add some description?

**Comment by bschnurr** [comment]
see https://github.com/microsoft/pylance-release/issues/2533#issuecomment-1089268739

**Comment by bschnurr** [comment]
looks like there is still and issue possibly due do to `commands` being a folder inside the `ext` namespace

![image](https://user-images.githubusercontent.com/1946977/162841698-2dfc51bf-eef7-4e6a-8497-a5d1eedf6865.png)

**Comment by davidonlaptop** [comment]
YMMV but it seems that the default indexing isn't deep enough for auto-import to work when the function or class you import is not at the root level, so here's what I had to put in my settings.json for it to work properly:

```json
{
  "python.analysis.autoImportCompletions": true,
  "python.analysis.indexing": true,
  "python.analysis.packageIndexDepths": [
    {
      "name": "",
      "depth": 3
    },
    
    ]
}
```
